src/peer-store/README.md

```diff
@@ -1,3 +1,156 @@
+
+On different context, it is also possible to gather relevant information for the peers. For instance, in `dht` operations, nodes can exchanges data of peers they know as part of the `dht` operation. In this case, we can get information from a peer that we already know. As a consequence, the `PeerStore` should act accordingly and not replace the data it owns, but just try to merge it the discovered data is new. For example, discovered a new address of a peer.
+
+#### Act
+
+When the `PeerStore` data is updated, this information might be important for different parties.
+
+`js-libp2p` keeps a topology of peers for each protocol a node is running. This way, once a protocol is supported for a peer, the topology of that protocol should be informed that a new peer may be used and the subsystem can decide if it should open a new stream it that peer or not.
+
+Every time a peer needs to dial another peer, it is essential that it knows the multiaddrs used by the peer, in order to perform a successful dial to a peer. The same is true for pinging a peer.
+
+## PeerStore implementation
+
+(Note: except for highlighting the APIs functionallity, they should be better formally described on `API.md` file)
+
+#### API:
+
+Access to its underlying books:
+
+- `peerStore.protoBook.*`
+- `peerStore.addressBook.*`
+
+High level operations:
+
+- `peerStore.set(peerId, data, options)` or `events`
+
+High level set which should be able to identify the type of data received and forward to the appropriate book sets. More than a bridge, this aims to allow the combination of multiple data storage as follows:
+
+`data = { multiaddrs: [...], protocols: [...] }`
+
+---- (should be removed / re-written, but important for design decisions)
+
+One aspect that we need to consider is wether we should add information to every book, even if we don't have any relevant information for it. For instance, if we just discover a peer via a typical discovery service, we will have the `peerId` and an array of `multiaddr`. When we do `peerStore.set()`, should we also do `protoBook.set()` with an empty list of protocols? I don't see any advantage on adding to the remaining ones.
+
+**IMPORTANT:** This is one of the biggest design decisions to make (set vs events). The programmatic API is the easiest solution but it can provide users an API that they sould not need. If we go on an event-based approach, the `peerStore` should receive all the relevant subsystems (discovery, identifyService, ...) and sense the environment (via events) to gather the information that would need to be sent via the API. Thile the latest seems the best solution, it is the more complex one to implement, as we would ideally have an interface that those subsystems would need to implement and each time we have a new subsystem that needs to add data to the peerStore, we might need to update the `peer-store` codebase (or have a good set of abstractions).
```

Comment:
The thought behind consuming events was to not allow libp2p users to put data directly in the `PeerStore`, as only libp2p should be responsible for it afaik. This was also where the private properties were getting in, so that users would not be incentivised to change the books and eventually lead to non consistent data.
 But I totally agree that it seems over complex and unecessary

Reply:
drop this?

Reply:
Yeah I'd drop it and stick with an api approach.